File: Python/dbinfo.py
import sys
import sqlite3
from tableinfo import TableInfo

class DbInfo(object):
    def __init__(self, name):
        self.name = name
        self.conn = sqlite3.connect(name)
        self.tables = {}
        self.conn.text_factory = lambda x: str(x, 'utf-8', 'ignore')
        cursor = self.conn.cursor()
        cursor.execute("SELECT name FROM sqlite_master WHERE type =\'table\' AND name NOT LIKE \'sqlite_%\'")
        rows = cursor.fetchall()
        if len(rows) > 0:
            tableNames = []
            for row in rows:
                tableNames.append(row[0])

            for tableName in tableNames:
                self.tables[tableName] = TableInfo(self.conn, tableName)

    def compareTables(self, tableName, numColumns, db):
        cursor = self.conn.cursor()
        sql = "SELECT * FROM '" + tableName + "'"
        cursor.execute(sql)
        rows = cursor.fetchall()
        if len(rows) > 0:
            cursor2 = db.conn.cursor()
            cursor2.execute(sql)

            for rowNum, row in enumerate(rows):
                row2 = cursor2.fetchone()
                if (row is None) or (row2 is None):
                    return False, rowNum
                for col in range(numColumns):
                    if row[col] != row2[col]:
                        return False, rowNum

            return True, 0

        return False, 0

    def compare(self, db):
        matches = True
        matched = []
        onlyOne = []
        onlyTwo = []
        for tableName, tableInfo in self.tables.items():
            tableInfo2 = db.tables.get(tableName)
            if tableInfo2 is not None:
                if tableInfo.compare(tableInfo2):
                    if tableInfo.numRows < 1000:
                        dataMatched, rowNum = self.compareTables(tableName, len(tableInfo.columns), db)
                        if not dataMatched:
                            matches = False
                            sys.stdout.write('Different Data for Table: {} in row {}\n'.format(tableName, rowNum + 1))
                        else:
                            matched.append(tableInfo)
                    else:
                        matched.append(tableInfo)
                else:
                    matches = False
                    sys.stdout.write('Different Table: {} {} {} but in {} {}\n'.format(tableName, self.name, tableInfo.toStr(False), 
                        db.name, tableInfo2.toStr(False)))
            else:
                matches = False
                onlyOne.append(tableName)

        for tableName, tableInfo in db.tables.items():
            if tableName not in self.tables:
                matches = False
                onlyTwo.append(tableName)

        if len(matched) > 0:
            sys.stdout.write("*************** {} matched tables ****************\n".format(len(matched)))
            for table in matched:
                sys.stdout.write("Table: {}\n".format(table.toStr(True)))

        if len(onlyOne) > 0:
            sys.stdout.write("*************** {} tables only in {} ****************\n".format(len(onlyOne), self.name))
            for table in onlyOne:
                sys.stdout.write("Table: {}\n".format(table))

        if len(onlyTwo) > 0:
            sys.stdout.write("*************** {} tables only in {} ****************\n".format(len(onlyTwo), db.name))
            for table in onlyTwo:
                sys.stdout.write("Table: {}\n".format(table))

        return matches

File: C#/DbInfo.cs
using Microsoft.Data.Sqlite;
using System;
using System.Collections.Generic;
using System.Data.Common;
using System.Text;

namespace dbComp
{
    class DbInfo
    {
        String name;
        Dictionary<String, TableInfo> tables = new Dictionary<String, TableInfo>();
        DbConnection sqlconn;
        public DbInfo(String path)
        {
            this.name = path;
            this.sqlconn = new SqliteConnection(String.Format("Data Source={0}", path));
            sqlconn.Open();

            var command = sqlconn.CreateCommand();
            command.CommandText = "SELECT name FROM sqlite_master WHERE type =\'table\' AND name NOT LIKE \'sqlite_%\'";
            using var reader = command.ExecuteReader();

            List<string> tableNames = new List<string>();
            while (reader.Read())
            {
                tableNames.Add(reader.GetString(0));
            }
            foreach (string tableName in tableNames)
            {
                tables.Add(tableName, new TableInfo(sqlconn, tableName));
            }
       }
        Tuple<bool, int> CompareTables(String tableName, DbInfo db)
        {
            var command = sqlconn.CreateCommand();
            command.CommandText = "SELECT * FROM '" + tableName + "'";
            using var reader = command.ExecuteReader();
            var command2 = db.sqlconn.CreateCommand();
            command2.CommandText = "SELECT * FROM '" + tableName + "'";
            using var reader2 = command2.ExecuteReader();

            Object[] values = new Object[reader.FieldCount];
            Object[] values2 = new Object[reader2.FieldCount];
            int nRow = 0;
            while (reader.Read())
            {
                reader2.Read();
                reader.GetValues(values);
                reader2.GetValues(values2);
                for (int nCol = 0; nCol < reader.FieldCount; ++nCol)
                {
                    if (!values[nCol].Equals(values2[nCol]))
                        return new Tuple<bool, int>(false, nRow);
                }
                ++nRow;
            }
            return new Tuple<bool, int>(true, 0);
        }
        public bool Compare(DbInfo db, ref List<TableInfo> tablesIn1, ref List<TableInfo> tablesIn2,
                        ref List<Tuple<TableInfo, TableInfo>> tablesMatched, ref List<Tuple<TableInfo, TableInfo, int>> tablesUnmatched)
        {
            foreach (var table in tables)
            {
                if (db.tables.ContainsKey(table.Key))
                {
                    var tableInfo = db.tables[table.Key];
                    if (table.Value.Compare(tableInfo))
                    {
                        if (tableInfo.NumRows < 1000)
                        {
                            var result = CompareTables(table.Key, db);
                            if (result.Item1)
                                tablesMatched.Add(new Tuple<TableInfo, TableInfo>(table.Value, tableInfo));
                            else
                                tablesUnmatched.Add(new Tuple<TableInfo, TableInfo, int>(table.Value, tableInfo, result.Item2));
                        }
                    }
                    else
                        tablesUnmatched.Add(new Tuple<TableInfo, TableInfo, int>(table.Value, tableInfo, -1));
                }
                else
                    tablesIn1.Add(table.Value);
            }
            foreach (var table in db.tables)
            {
                if (!tables.ContainsKey(table.Key))
                    tablesIn2.Add(table.Value);
            }
            return (tablesIn1.Count == 0) && (tablesIn2.Count == 0) && (tablesUnmatched.Count == 0);
        }
    }
}
